### PR TITLE
[Scenes] Scene Handler Encode and Decode attribute value list 

### PR DESCRIPTION
**Pull request by lpbeliveau-silabs**
Helper functions for SceneHandlers to avoid code duplication went implementing cluster specific handlers.

### CODE REVIEW

File: src/app/clusters/scenes-server/SceneTableImpl.h [hunk 1]
```diff
@@ -46,89 +46,110 @@ class DefaultSceneHandlerImpl : public scenes::SceneHandler
...
 
+    /// @brief Encodes an attribute value list into a TLV structure and resizes the buffer to the size of the encoded data
+    /// @param aVlist[in] Attribute value list to encode
+    /// @param serializedBytes[out] Buffer to fill from the Attribute value list in a TLV format
+    /// @return CHIP_ERROR
+    virtual CHIP_ERROR
+    EncodeAttributeValueList(const app::DataModel::List<app::Clusters::Scenes::Structs::AttributeValuePair::Type> & aVlist,
+                             MutableByteSpan & serializedBytes)
+    {
+        TLV::TLVWriter writer;
```
Comment (bzbarsky-apple) [inline src/app/clusters/scenes-server/SceneTableImpl.h:57]: Why is the implementation inlined here?  This is virtual anyway...  Please put the implementation in the .cpp?

File: src/app/clusters/scenes-server/SceneTableImpl.h [hunk 2]
```diff
@@ -46,89 +46,110 @@ class DefaultSceneHandlerImpl : public scenes::SceneHandler
...
+        TLV::TLVType outer;
+        writer.Init(serializedBytes);
+        ReturnErrorOnFailure(writer.StartContainer(TLV::AnonymousTag(), TLV::kTLVType_Structure, outer));
```
Comment (bzbarsky-apple) [inline src/app/clusters/scenes-server/SceneTableImpl.h:60]: I have to ask: why is this outer structure envelope needed?  Is the idea to allow extensibility later?  It might be worth documenting this part.

File: src/app/clusters/scenes-server/SceneTableImpl.h [hunk 3]
```diff
@@ -46,89 +46,110 @@ class DefaultSceneHandlerImpl : public scenes::SceneHandler
...
+        TLV::TLVReader reader;
+        TLV::TLVType outer;
+        size_t pairTotal = 0;
+
+        reader.Init(serializedBytes);
+        ReturnErrorOnFailure(reader.Next(TLV::kTLVType_Structure, TLV::AnonymousTag()));
+        ReturnErrorOnFailure(reader.EnterContainer(outer));
+        ReturnErrorOnFailure(reader.Next(
+            TLV::kTLVType_Array, TLV::ContextTag(app::Clusters::Scenes::Structs::ExtensionFieldSet::Fields::kAttributeValueList)));
+        aVlist.Decode(reader);
```
Comment (bzbarsky-apple) [inline src/app/clusters/scenes-server/SceneTableImpl.h:86]: Does this not need an error check?

File: src/app/clusters/scenes-server/SceneTableImpl.h [hunk 4]
```diff
@@ -46,89 +46,110 @@ class DefaultSceneHandlerImpl : public scenes::SceneHandler
...
+        ReturnErrorOnFailure(reader.ExitContainer(outer));
+        // Verify size of list
+        ReturnErrorOnFailure(aVlist.ComputeSize(&pairTotal));
+        VerifyOrReturnError(pairTotal <= ArraySize(mAVPairs), CHIP_ERROR_BUFFER_TOO_SMALL);
```
Comment (bzbarsky-apple) [inline src/app/clusters/scenes-server/SceneTableImpl.h:90]: What does this function have to do with mAVPairs?

File: src/app/clusters/scenes-server/SceneTableImpl.h [hunk 5]
```diff
@@ -46,89 +46,110 @@ class DefaultSceneHandlerImpl : public scenes::SceneHandler
...
+            app::Clusters::Scenes::Structs::AttributeValuePair::DecodableType aVPair = pair_iterator.GetValue();
+            aVPairs[pairCount].attributeID                                           = aVPair.attributeID;
+            aVPairs[pairCount].attributeValue                                        = aVPair.attributeValue;
             pairCount++;
         }
         ReturnErrorOnFailure(pair_iterator.GetStatus());
-
-        app::DataModel::List<app::Clusters::Scenes::Structs::AttributeValuePair::Type> attributeValueList;
-        attributeValueList = mAVPairs;
+        app::DataModel::List<app::Clusters::Scenes::Structs::AttributeValuePair::Type> attributeValueList(aVPairs);
```
Comment (bzbarsky-apple) [inline src/app/clusters/scenes-server/SceneTableImpl.h:122]: A lot of this code could benefit from some judicious "using" in this class scope... ;) Followup for that?

File: src/app/clusters/scenes-server/SceneTableImpl.h [hunk 6]
```diff
@@ -46,89 +46,110 @@ class DefaultSceneHandlerImpl : public scenes::SceneHandler
...
 
-        writer.Init(serializedBytes);
-        ReturnErrorOnFailure(writer.StartContainer(TLV::AnonymousTag(), TLV::kTLVType_Structure, outer));
-        ReturnErrorOnFailure(app::DataModel::Encode(
-            writer, TLV::ContextTag(app::Clusters::Scenes::Structs::ExtensionFieldSet::Fields::kAttributeValueList),
-            attributeValueList));
-        ReturnErrorOnFailure(writer.EndContainer(outer));
-
-        return CHIP_NO_ERROR;
+        return EncodeAttributeValueList(attributeValueList, serializedBytes);
```
Comment (bzbarsky-apple) [inline src/app/clusters/scenes-server/SceneTableImpl.h:125]: So this used to not update the size of `serializedBytes`, right?  Was that just a bug?

File: src/app/clusters/scenes-server/SceneTableImpl.h [hunk 7]
```diff
@@ -46,89 +46,110 @@ class DefaultSceneHandlerImpl : public scenes::SceneHandler
...
-        ReturnErrorOnFailure(reader.Next(
-            TLV::kTLVType_Array, TLV::ContextTag(app::Clusters::Scenes::Structs::ExtensionFieldSet::Fields::kAttributeValueList)));
-        attributeValueList.Decode(reader);
-
-        // Verify size of list
-        ReturnErrorOnFailure(attributeValueList.ComputeSize(&pairTotal));
-        VerifyOrReturnError(pairTotal <= ArraySize(mAVPairs), CHIP_ERROR_BUFFER_TOO_SMALL);
+        ReturnErrorOnFailure(DecodeAttributeValueList(serializedBytes, attributeValueList));
 
+        uint8_t pairCount  = 0;
```
Comment (bzbarsky-apple) [inline src/app/clusters/scenes-server/SceneTableImpl.h:141]: Isn't _this_ where the count check against mAVPairs should happen, since this is where we care about it?

File: src/app/clusters/scenes-server/SceneTableImpl.h [hunk 8]
```diff
@@ -46,89 +46,110 @@ class DefaultSceneHandlerImpl : public scenes::SceneHandler
...
     virtual CHIP_ERROR SerializeAdd(EndpointId endpoint,
                                     const app::Clusters::Scenes::Structs::ExtensionFieldSet::DecodableType & extensionFieldSet,
                                     MutableByteSpan & serializedBytes) override
     {
-        app::Clusters::Scenes::Structs::AttributeValuePair::DecodableType aVPair;
-        TLV::TLVWriter writer;
-        TLV::TLVType outer;
-        size_t pairTotal  = 0;
-        uint8_t pairCount = 0;
+        app::Clusters::Scenes::Structs::AttributeValuePair::Type aVPairs[kMaxAvPair];
```
Comment (bzbarsky-apple) [inline src/app/clusters/scenes-server/SceneTableImpl.h:105]: Why the change to not use mAVPairs here?

Comment (lpbeliveau-silabs) [reply src/app/clusters/scenes-server/SceneTableImpl.h:105]: mAVPairs was originally introduced to ensure ExtensionFieldSet::Type.attributeValueList's pointer stays allocated after DefaultSceneHandlerImpl::Deserialize() returns. Since the data from aVPairs is all serialized into a buffer, I figured there is no need to share the mAVPairs buffer.

File: src/app/clusters/scenes-server/SceneTableImpl.h (hunk 6 above)
Comment (lpbeliveau-silabs) [reply src/app/clusters/scenes-server/SceneTableImpl.h:125]: Yes, that was a bug that led each EFS to always take the maximum allowed size.

File: src/app/clusters/scenes-server/SceneTableImpl.h (hunk 2 above)
Comment (lpbeliveau-silabs) [reply src/app/clusters/scenes-server/SceneTableImpl.h:60]: This (and the resulting failure when trying to encode without it) was my interpretation of Matter main spec:

> 5.1.5. TLV Content
> [...]All elements SHALL be housed within an anonymous top-level structure container.

Then in List.h, the encode function will encode a TLV::kTLVType_Array, which, when associated with an TLV::AnonymousTag(), caused WriteElementHead to fail.

So it lead me to believe than one cannot encode a List without an outer structure container.

File: src/app/clusters/scenes-server/SceneTableImpl.h (hunk 5 above)
Comment (lpbeliveau-silabs) [reply src/app/clusters/scenes-server/SceneTableImpl.h:122]: Issue logged [here](https://github.com/project-chip/connectedhomeip/issues/27111)

File: src/app/clusters/scenes-server/SceneTableImpl.h (hunk 2 above)
Comment (bzbarsky-apple) [reply src/app/clusters/scenes-server/SceneTableImpl.h]: 5.1.5 is about the TLV parts of a commissioning payload.  That's not a general TLV requirement.

> The encode function will encode a TLV::kTLVType_Array, which, when associated with an TLV::AnonymousTag(), caused WriteElementHead to fail

That's odd.  I would expect this to work fine....  If you look at `TLVPacketBufferBackingStoreTest::BasicEncodeDecode` is uses a toplevel anonymous-tagged TLV array for that test, and it works fine.

Comment (lpbeliveau-silabs) [reply src/app/clusters/scenes-server/SceneTableImpl.h]: Indeed, I think that I must have tried to keep the data model type `app::Clusters::Scenes::Structs::ExtensionFieldSet::Fields::kAttributeValueList` for some obscure reason.

Regardless, it works fine without it so I removed it.

File: src/app/clusters/scenes-server/SceneTableImpl.h [hunk 9]
```diff
@@ -46,94 +46,40 @@ class DefaultSceneHandlerImpl : public scenes::SceneHandler
...
+    /// @param aVlist[in] Attribute value list to encode
+    /// @param serializedBytes[out] Buffer to fill from the Attribute value list in a TLV format
+    /// @return CHIP_ERROR
+    virtual CHIP_ERROR
+    EncodeAttributeValueList(const app::DataModel::List<app::Clusters::Scenes::Structs::AttributeValuePair::Type> & aVlist,
+                             MutableByteSpan & serializedBytes);
+
+    /// @brief Decodes an attribute value list from a TLV structure and ensure it fits the member pair buffer
+    /// @param serializedBytes [in] Buffer to read from
+    /// @param aVlist [out] Attribute value list to fill from the TLV structure
```
Comment (bzbarsky-apple) [inline src/app/clusters/scenes-server/SceneTableImpl.h:59]:
```suggestion
    /// @param aVlist [out] Attribute value list to fill from the TLV structure.  Only valid while the buffer backing serializedBytes exists and its contents are not modified.
```

File: src/app/clusters/scenes-server/SceneTableImpl.cpp [hunk 10]
```diff
@@ -22,6 +22,97 @@
...
+{
+    TLV::TLVReader reader;
+    // TLV::TLVType outer;
+
+    reader.Init(serializedBytes);
+    // ReturnErrorOnFailure(reader.Next(TLV::kTLVType_Structure, TLV::AnonymousTag());
+    // ReturnErrorOnFailure(reader.EnterContainer(outer));
+    ReturnErrorOnFailure(reader.Next(TLV::kTLVType_Array, TLV::AnonymousTag()));
+    ReturnErrorOnFailure(aVlist.Decode(reader));
+    // ReturnErrorOnFailure(reader.ExitContainer(outer));
```
Comment (bzbarsky-apple) [inline src/app/clusters/scenes-server/SceneTableImpl.cpp:53]:
```suggestion

    reader.Init(serializedBytes);
    ReturnErrorOnFailure(reader.Next(TLV::kTLVType_Array, TLV::AnonymousTag()));
    ReturnErrorOnFailure(aVlist.Decode(reader));
```

File: src/app/clusters/scenes-server/SceneTableImpl.cpp [hunk 11]
```diff
@@ -22,6 +22,97 @@
...
+    while (pair_iterator.Next())
+    {
+        app::Clusters::Scenes::Structs::AttributeValuePair::DecodableType aVPair = pair_iterator.GetValue();
+        aVPairs[pairCount].attributeID                                           = aVPair.attributeID;
+        aVPairs[pairCount].attributeValue                                        = aVPair.attributeValue;
+        pairCount++;
+    }
+    ReturnErrorOnFailure(pair_iterator.GetStatus());
+    app::DataModel::List<app::Clusters::Scenes::Structs::AttributeValuePair::Type> attributeValueList(aVPairs);
+    attributeValueList.reduce_size(pairCount);
```
Comment (bzbarsky-apple) [inline src/app/clusters/scenes-server/SceneTableImpl.cpp:81]:
```suggestion
    app::DataModel::List<app::Clusters::Scenes::Structs::AttributeValuePair::Type> attributeValueList(aVPairs, pairCount);
```

File: src/app/clusters/scenes-server/SceneTableImpl.cpp [hunk 12]
```diff
@@ -22,6 +22,97 @@
...
+    ReturnErrorOnFailure(attributeValueList.ComputeSize(&pairTotal));
+    VerifyOrReturnError(pairTotal <= ArraySize(mAVPairs), CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    uint8_t pairCount  = 0;
+    auto pair_iterator = attributeValueList.begin();
+    while (pair_iterator.Next())
+    {
+        app::Clusters::Scenes::Structs::AttributeValuePair::DecodableType decodePair = pair_iterator.GetValue();
+        mAVPairs[pairCount].attributeID                                              = decodePair.attributeID;
+        mAVPairs[pairCount].attributeValue                                           = decodePair.attributeValue;
```
Comment (bzbarsky-apple) [inline src/app/clusters/scenes-server/SceneTableImpl.cpp:104]: Again, any reason not to do:
```suggestion
        mAVPairs[pairCount] = pair_iterator.GetValue();
```
?

**Review by bzbarsky-apple — APPROVED** [review]

**Review by jmartinez-silabs — APPROVED** [review]
